lazyflow/utility/roiRequestBuffer.py

```diff
@@ -0,0 +1,98 @@
+from typing import Any, Dict, Iterator, Tuple, Type
+from queue import Queue
+
+import numpy
+
+from lazyflow.graph import Slot
+from lazyflow.utility.roiRequestBatch import RoiRequestBatch
+
+
+ROI = Tuple[int, int, int, int, int]
+SHAPE = ROI
+ROI_TUPLE = Tuple[ROI, ROI]
+
+
+class _RoiIter:
+    """Iterate over rois for a given shape"""
+
+    def __init__(self, shape: SHAPE):
+        self._shape = shape
+
+    def to_index(self, roi: ROI_TUPLE) -> int:
+        """Convert ROI_TUPLE to index
+
+        first ROI_TUPLE -> index = 0
+        last ROI_TUPLE -> index = len(self) - 1
+        """
+        return numpy.ravel_multi_index(roi[0][:-2], self._shape[:-2])
+
+    def __len__(self) -> int:
+        return numpy.prod(self._shape[:-2])
+
+    def __iter__(self) -> Iterator[ROI_TUPLE]:
+        """iterate in ascending order, according to to_index"""
+        for partial_start in numpy.ndindex(self._shape[:-2]):
+            start: ROI = partial_start + (0, 0)
+            stop: ROI = tuple(i + 1 for i in partial_start) + self._shape[-2:]
+            yield start, stop
+
+
+class RoiRequestBuffer:
+    """Iterate over first 3 dims of a 5D slot in nd-ascending order."""
+
+    def __init__(self, slot: Slot, batchsize: int, roi_iter_cls: Type[_RoiIter] = _RoiIter):
+        """
+        Args:
+            slot: slot to request data from
+            batchsize: maximum number of requests to launch in parallel
+            roi_iter_cls: mainly exposed for testing, class that produces iterators\
+            given a shape
+        """
+        self._slot = slot
+        self._batchsize = batchsize
+
+        self._items: Dict[int, numpy.ndarray] = {}
+        self._q: Queue[Tuple[ROI_TUPLE, numpy.ndarray]] = Queue()
+        self._index = 0
+        self._roi_iter = roi_iter_cls(self._slot.meta.shape)
+        self._max = len(self._roi_iter)
+
+        self._roi_request_batch = self._config_batch_request()
+
+    def _config_batch_request(self) -> RoiRequestBatch:
```

Comment:
hmmm readability? :D

This stems from the times I thought that it could make sense to make this thing iterate multiple times over the data. So before I had it in `__iter__`... but yeah, one could probably shove it into `__init__`